docs/source/format/CDataInterface/PyCapsuleInterface.rst

```diff
@@ -0,0 +1,238 @@
+the data and marked the release callback as null, so there isn’t a risk of
+releasing data the consumer is using. Read more in the C Data Interface
+specification.
+
+Because of the lifetime semantics of the underlying Arrow structs, the
+PyCapsules can only be used once.
+
+
+Export Protocol
+===============
+
+The interface is three separate protocols:
+
+* ``ArrowSchemaExportable``, which defines the ``__arrow_c_schema__`` method.
+* ``ArrowArrayExportable``, which defines the ``__arrow_c_array__``. It extends
+  ``ArrowSchemaExportable``, so it must also define ``__arrow_c_schema__``.
+* ``ArrowStreamExportable``, which defines the ``__arrow_c_stream__``. It extends
+  ``ArrowSchemaExportable``, so it must also define ``__arrow_c_schema__``.
+
+The protocols are defined below in terms of ``typing.Protocol``. These may be
+copied into a library for the purposes of static type checking, but this is not
+required to implement the protocol.
+
+.. TODO: should stream have a batch size parameter?
+.. TODO: should we have a allow_copy flag?
+
+
+
+.. code-block:: python
+
+    from typing import Tuple, Protocol
+    from typing_extensions import Self
+
+    class ArrowSchemaExportable(Protocol):
+        def __arrow_c_schema__(self) -> object:
+            """
+            Get a PyCapsule containing a C ArrowSchema representation of the object.
+
+            The capsule will have a name of "arrowschema".
+            """
+            ...
+
+    class ArrowArrayExportable(ArrowSchemaExportable, Protocol):
+        def __arrow_c_array__(
+            self,
+            requested_schema: object | None = None
+        ) -> Tuple[object, object]:
+            """
+            Export array as a pair of PyCapsules for the ArrowSchema and ArrowArray.
+
+            The capsule will have a name of "arrowarray".
+
+            If requested_schema is passed, the callee should attempt to provide the
+            data in the requested schema. However, this is best-effort, and the
+            callee may return a PyCapsule containing an ArrowArray with a different
+            schema. This parameter is useful for cases where the underlying data
+            could be represented in multiple ways, and the caller has a preference
+            for how it is represented. For example, some systems have a single
+            integer type, but Arrow has multiple integer types with different 
+            sizes and sign.
+
+            Parameters
+            ----------
+            requested_schema : PyCapsule | None
+                A PyCapsule containing a C ArrowSchema representation of a requested
+                schema. Conversion to this schema is best-effort.
+
+            Returns
+            -------
+            Tuple[PyCapsule, PyCapsule]
+                A pair of PyCapsules containing a C ArrowSchema and ArrowArray,
+                respectively.
+            """
+            ...
+
+
+    class ArrowStreamExportable(ArrowSchemaExportable, Protocol):
```

Comment:
I was thinking any object that can be exported as an array or stream also has a schema that can be exported. And I think it can be useful because just exporting the schema tells the caller what schema the data would be exported as by default if no `requested_schema` was provided.

Reply:
> And for ArrowArrayExportable, we also want to expose the schema directly, I think, so this keeps it consistent

This means you could eventually pass an array or record batch to `pa.schema`, which seems weird.

Reply:
In theory our functions like `pa.schema(..)` could reject objects that also implement `__arrow_c_array/stream__` and not only the schema one, if we want so. 

Personally I think we definitely want to give the ability to inspect the schema of an array object, before asking for the actual data (as this step could in certain cases not be zero-copy, and thus have a cost). For pyarrow objects you could check the `.type` or `.schema` attributes, and then get `__arrow_c_schema__`, but that gives again something library-specific in the middle, which we want to avoid.

Reply:
> This means you could eventually pass an array or record batch to pa.schema, which seems weird.

I see your point. That would be weird indeed.

I can remove this inheritance, and I'll think about whether we actually need a method to get a schema on stream / array exportables.

Reply:
> Personally I think we definitely want to give the ability to inspect the schema of an array object, before asking for the actual data (as this step could in certain cases not be zero-copy, and thus have a cost).

When would this not be zero-copy? Also, an array object wouldn't have a schema ever anyways, right - it would be a separate object (or an object implementing both protocols, I suppose)

I agree we should treat them as separate objects. Note that getting the schema of a stream may involve I/O and is a falliable operation, so I think that's more reason to separate them.

Reply:
> When would this not be zero-copy? 

For any object that does not use Arrow memory _exactly_. For example data from Velox or DuckDB right now, or pandas DataFrames, .. 
A general question might be to what extent a library should add those dunder methods if those involve a conversion step (or whether there is a general expectation that those are guaranteed to be zero-copy)

Reply:
Ah, okay. Those systems could do the conversion lazily (only when you call get_next), but that's fair. (But you still don't want streams to implicitly convert to schemas, since then a system that does the conversion eagerly would do the work when you just want a schema.)

Reply:
FYI I've removed the inheritance here.

Reply:
This is also something that could be handled by a protocol "higher up", like the dataset protocol we have been discussing (https://github.com/apache/arrow/issues/37504). So that way, an object could provide both a schema and a stream access, and thus a way to first inspect the schema before asking the actual stream.